lib/system.nim

```diff
@@ -2410,42 +2410,52 @@ proc `$`*[T: tuple|object](x: T): string =
       result.add("...")
   result.add(")")
 
-proc collectionToString[T: set | seq](x: T, b, e: string): string =
-  when x is seq:
-    if x.isNil: return "nil"
-  result = b
+proc collectionToString[T](x: T, prefix, separator, suffix: string): string =
+  result = prefix
   var firstElement = true
   for value in items(x):
-    if not firstElement: result.add(", ")
+    if firstElement:
+      firstElement = false
+    else:
+      result.add(separator)
+
     when compiles(value.isNil):
-      if value.isNil: result.add "nil"
-      else: result.add($value)
+      # this branch should not be necessary
+      if value.isNil:
+        result.add "nil"
+      else:
+        result.add($value)
+    # prevent temporary string allocation
+    elif compiles(result.add(value)):
+      # don't insert '\0' characters into the result string
+      when value is char:
```

Comment:
Well The reason I put it in there is this:

```
type
  MyType = object
    x: array[10,char]
    a,b,c: int
    s: string
var mt: MyType
echo mt
```
The result string will just end at the first character of the array. I don't think that is very usable, therefore I decided to simply not add null characters to the string. And the string that is created now is reversable. Whenever there are two separators in the result string without a character in between, the character was a null character. So you can think of it like the null character is converted into the empty string. So I think your point that this `$` operator would only be usable for `echo` is not quite true. But if you really don't like this behavior, I can change it like you have. I just think an `$` operator that returns "[a,b,c,,,,,,,]" on an `array[10,char]` is more useful than an operator that returns "[a,b,c,"